src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,282 @@
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            int? leftCardinality = GetTupleCardinality(left);
+            int? rightCardinality = GetTupleCardinality(right);
+
+            if (leftCardinality.HasValue && rightCardinality.HasValue)
+            {
+                TypeSymbol leftType = left.Type;
+                TypeSymbol rightType = right.Type;
+
+                if (leftCardinality.Value != rightCardinality.Value)
+                {
+                    Error(diagnostics, ErrorCode.ERR_TupleSizesMismatchForBinOps, node, leftCardinality, rightCardinality);
+
+                    return new TupleBinaryOperatorInfo.Single(leftType ?? CreateErrorType(), rightType ?? CreateErrorType(),
+                        BinaryOperatorKind.Error, methodSymbolOpt: null, boolConversion: Conversion.NoConversion, boolOperator: default);
+                }
+
+                // typeless tuple literals are not nullable
+                bool leftNullable = leftType?.IsNullableType() == true;
+                bool rightNullable = rightType?.IsNullableType() == true;
+
+                return BindTupleBinaryOperatorNestedInfo(node, kind,
+                    GetTupleArgumentsOrPlaceholders(left), GetTupleArgumentsOrPlaceholders(right),
+                    leftNullable || rightNullable, diagnostics);
+            }
+
+            return BindTupleBinaryOperatorSingleInfo(node, kind, left, right, diagnostics);
+        }
+
+        private TupleBinaryOperatorInfo BindTupleBinaryOperatorSingleInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            TypeSymbol leftType = left.Type;
+            TypeSymbol rightType = right.Type;
+
+            if ((object)leftType != null && leftType.IsDynamic() || (object)rightType != null && rightType.IsDynamic())
+            {
+                return BindTupleDynamicBinaryOperatorSingleInfo(node, kind, left, right, diagnostics);
+            }
+
+            LookupResultKind resultKind;
+            BinaryOperatorAnalysisResult best = this.BinaryOperatorOverloadResolution(kind, left, right, node, diagnostics, out resultKind, out var _);
+
+            MethodSymbol resultMethod = null;
+            BinaryOperatorKind resultOperatorKind;
+            TypeSymbol convertedLeftType;
+            TypeSymbol convertedRightType;
+            TypeSymbol returnType;
+            bool hasErrors;
+
+            if (!best.HasValue)
+            {
+                resultOperatorKind = kind;
+                convertedLeftType = leftType ?? CreateErrorType();
+                convertedRightType = rightType ?? CreateErrorType();
+                returnType = CreateErrorType();
+                hasErrors = true;
+            }
+            else
+            {
+                bool leftNull = left.IsLiteralNull();
```

Comment:
>leftNull = left.IsLiteralNull(); [](start = 21, length = 32)

Should default literals be handled specially too?

Reply:
It will (tracked in [https://github.com/dotnet/roslyn/issues/22937](https://github.com/dotnet/roslyn/issues/22937)), but that will be a subsequent PR because I need to flow the fix for `default` into this feature branch first.

---
In reply to: [170492819](https://github.com/dotnet/roslyn/pull/23356#discussion_r170492819) [](ancestors = 170492819)